File: src/normalize.js
const { SpecError } = require('./errors');

const HTTP_METHODS = ['get', 'put', 'post', 'delete', 'options', 'head', 'patch', 'trace'];
const STATUS_PATTERN = /^(default|[1-5]\d{2}|[1-5]XX)$/i;

function asArray(value) {
  if (value === undefined || value === null) return [];
  return Array.isArray(value) ? value : [value];
}

function contentMap(content = {}) {
  if (!content || typeof content !== 'object' || Array.isArray(content)) return {};
  return Object.fromEntries(Object.entries(content).map(([mediaType, value]) => [
    mediaType,
    {
      mediaType,
      schema: value && value.schema,
      example: value && value.example,
      examples: value && value.examples
    }
  ]));
}

function parameter(parameter) {
  if (!parameter || typeof parameter !== 'object') return null;
  return {
    name: parameter.name,
    in: parameter.in,
    required: Boolean(parameter.required),
    description: parameter.description,
    deprecated: Boolean(parameter.deprecated),
    style: parameter.style,
    explode: parameter.explode,
    schema: parameter.schema || (parameter.type ? {
      type: parameter.type,
      format: parameter.format,
      items: parameter.items,
      enum: parameter.enum,
      default: parameter.default
    } : undefined),
    example: parameter.example,
    examples: parameter.examples
  };
}

function normalizeParameters(pathParameters, operationParameters) {
  const merged = new Map();
  [...asArray(pathParameters), ...asArray(operationParameters)].forEach((value) => {
    const normalized = parameter(value);
    if (!normalized || !normalized.name || !normalized.in) return;
    merged.set(`${normalized.in}:${normalized.name}`, normalized);
  });
  return [...merged.values()];
}

function normalizeRequestBody(operation, root) {
  if (operation.requestBody && typeof operation.requestBody === 'object') {
    return {
      required: Boolean(operation.requestBody.required),
      description: operation.requestBody.description,
      content: contentMap(operation.requestBody.content)
    };
  }
  const bodyParameters = asArray(operation.parameters).filter((item) => item && item.in === 'body');
  const formParameters = asArray(operation.parameters).filter((item) => item && item.in === 'formData');
  if (!bodyParameters.length && !formParameters.length) return undefined;
  const consumes = asArray(operation.consumes || root.consumes || ['application/json']);
  const body = bodyParameters[0];
  const schema = body && body.schema;
  const properties = formParameters.reduce((result, item) => {
    result[item.name] = {
      type: item.type,
      format: item.format,
      items: item.items,
      description: item.description
    };
    return result;
  }, {});
  const required = formParameters.filter((item) => item.required).map((item) => item.name);
  return {
    required: Boolean((body && body.required) || required.length),
    description: body && body.description,
    content: Object.fromEntries(consumes.map((mediaType) => [mediaType, {
      mediaType,
      schema: schema || (formParameters.length ? { type: 'object', properties, required } : undefined)
    }]))
  };
}

function normalizeResponses(operation, root) {
  const responses = operation.responses;
  if (!responses || typeof responses !== 'object' || Array.isArray(responses)) {
    throw new SpecError('Each operation must define a responses object.', 'MISSING_RESPONSES');
  }
  const produces = asArray(operation.produces || root.produces);
  return Object.entries(responses).map(([statusCode, response]) => {
    if (!STATUS_PATTERN.test(statusCode)) {
      throw new SpecError(`Invalid response status "${statusCode}".`, 'INVALID_RESPONSE_STATUS');
    }
    const value = response && typeof response === 'object' ? response : {};
    let content = contentMap(value.content);
    if (!Object.keys(content).length && value.schema) {
      content = Object.fromEntries(produces.map((mediaType) => [mediaType, { mediaType, schema: value.schema }]));
    }
    return {
      statusCode,
      description: value.description || '',
      headers: value.headers || {},
      content
    };
  });
}

function normalizeSecurity(operation, root) {
  const security = operation.security !== undefined ? operation.security : root.security;
  return security === undefined ? [] : asArray(security);
}

function normalizeSecuritySchemes(root) {
  const schemes = root.components && root.components.securitySchemes
    ? root.components.securitySchemes
    : root.securityDefinitions || {};
  return Object.fromEntries(Object.entries(schemes).map(([name, scheme]) => [name, {
    name,
    type: scheme.type,
    scheme: scheme.scheme,
    bearerFormat: scheme.bearerFormat,
    in: scheme.in,
    parameterName: scheme.name,
    flows: scheme.flows,
    scopes: scheme.scopes || {}
  }]));
}

function normalizePath(path) {
  if (path === '/' || path === '') return '/';
  return `/${path.replace(/^\/+|\/+$/g, '')}`;
}

function inventory(spec) {
  if (!spec || typeof spec !== 'object' || Array.isArray(spec)) {
    throw new SpecError('The parsed specification must be an object.', 'INVALID_SPEC');
  }
  const version = spec.openapi || spec.swagger;
  if (typeof version !== 'string' || !(version.startsWith('3.') || version.startsWith('2.'))) {
    throw new SpecError('Only OpenAPI 3.x and Swagger 2.x specifications are supported.', 'UNSUPPORTED_VERSION');
  }
  if (!spec.info || typeof spec.info !== 'object') {
    throw new SpecError('Specification is missing required "info" metadata.', 'MISSING_INFO');
  }
  if (!spec.paths || typeof spec.paths !== 'object' || Array.isArray(spec.paths)) {
    throw new SpecError('Specification is missing a valid "paths" object.', 'MISSING_PATHS');
  }
  const endpoints = [];
  for (const [rawPath, pathItem] of Object.entries(spec.paths)) {
    if (!pathItem || typeof pathItem !== 'object') continue;
    for (const method of HTTP_METHODS) {
      const operation = pathItem[method];
      if (!operation) continue;
      if (typeof operation !== 'object' || Array.isArray(operation)) {
        throw new SpecError(`Operation ${method.toUpperCase()} ${rawPath} must be an object.`, 'INVALID_OPERATION');
      }
      const path = normalizePath(rawPath);
      const params = normalizeParameters(pathItem.parameters, operation.parameters);
      endpoints.push({
        id: operation.operationId || `${method}:${path}`,
        path,
        method: method.toUpperCase(),
        operationId: operation.operationId,
        summary: operation.summary,
        description: operation.description,
        deprecated: Boolean(operation.deprecated),
        tags: asArray(operation.tags),
        parameters: params,
        requestBody: normalizeRequestBody(operation, spec),
        responses: normalizeResponses(operation, spec),
        security: normalizeSecurity(operation, spec),
        servers: operation.servers || spec.servers
      });
    }
  }
  return {
    version,
    format: spec.openapi ? 'openapi' : 'swagger',
    info: { title: spec.info.title, version: spec.info.version, description: spec.info.description },
    servers: spec.servers || (spec.host ? [{ url: `${(spec.schemes || ['http'])[0]}://${spec.host}${spec.basePath || ''}` }] : []),
    securitySchemes: normalizeSecuritySchemes(spec),
    tags: asArray(spec.tags),
    endpoints
  };
}

module.exports = { inventory, normalizePath };

File: sentinel/normalize.py
"""Turn an OpenAPI 3.x / Swagger 2.0 dict into a normalized Inventory."""
from __future__ import annotations

from typing import Any

from .models import Endpoint, Inventory, Param

METHODS = ("get", "put", "post", "delete", "patch", "head", "options", "trace")
MAX_REF_DEPTH = 20


class RefResolver:
    """Resolves local '#/...' refs; stops on cycles/depth to stay safe."""

    def __init__(self, doc: dict[str, Any], warnings: list[str]):
        self.doc, self.warnings = doc, warnings

    def _lookup(self, ref: str) -> Any:
        if not ref.startswith("#/"):
            self.warnings.append(f"External $ref not resolved: {ref}")
            return {"$ref": ref}
        node: Any = self.doc
        for part in ref[2:].split("/"):
            part = part.replace("~1", "/").replace("~0", "~")
            if not isinstance(node, dict) or part not in node:
                self.warnings.append(f"Broken $ref: {ref}")
                return {"$ref": ref, "broken": True}
            node = node[part]
        return node

    def resolve(self, node: Any, seen: tuple[str, ...] = ()) -> Any:
        if isinstance(node, list):
            return [self.resolve(n, seen) for n in node]
        if not isinstance(node, dict):
            return node
        if "$ref" in node and isinstance(node["$ref"], str):
            ref = node["$ref"]
            if ref in seen or len(seen) >= MAX_REF_DEPTH:
                return {"$ref": ref, "circular": True}
            target = self._lookup(ref)
            if "$ref" in target and target.get("$ref") == ref:
                return target
            return self.resolve(target, seen + (ref,))
        return {k: self.resolve(v, seen) for k, v in node.items()}


def normalize(doc: dict[str, Any]) -> Inventory:
    warnings: list[str] = []
    r = RefResolver(doc, warnings)
    v2 = "swagger" in doc
    info = doc.get("info") or {}
    global_sec = doc.get("security", [])
    global_consumes = doc.get("consumes", ["application/json"])
    global_produces = doc.get("produces", ["application/json"])

    endpoints: list[Endpoint] = []
    for path, item in doc["paths"].items():
        item = r.resolve(item) if isinstance(item, dict) else {}
        shared = item.get("parameters", [])
        for method in METHODS:
            op = item.get(method)
            if not isinstance(op, dict):
                continue
            ep_id = f"{method.upper()} {path}"
            params, body = _params(shared, op.get("parameters", []), v2,
                                   op.get("consumes", global_consumes))
            if not v2:
                body = _body_v3(op.get("requestBody"))
            endpoints.append(Endpoint(
                id=ep_id,
                method=method.upper(),
                path=path,
                operation_id=op.get("operationId"),
                summary=op.get("summary") or op.get("description") or "",
                tags=op.get("tags", []),
                params=params,
                request_body=body,
                responses=_responses(op.get("responses", {}), v2,
                                     op.get("produces", global_produces)),
                security=op.get("security", global_sec) or [],
                deprecated=bool(op.get("deprecated")),
            ))
    if not endpoints:
        warnings.append("Spec contains no operations")

    return Inventory(
        title=info.get("title", "Untitled"),
        version=str(info.get("version", "")),
        spec_version="2.0" if v2 else str(doc["openapi"]),
        servers=_servers(doc, v2),
        security_schemes=r.resolve(
            doc.get("securityDefinitions", {}) if v2
            else (doc.get("components") or {}).get("securitySchemes", {})),
        endpoints=endpoints,
        warnings=sorted(set(warnings)),
    )


def _params(shared, own, v2, consumes):
    merged = {(p.get("name"), p.get("in")): p for p in shared + own if isinstance(p, dict)}
    params, body = [], None
    form_props: dict[str, Any] = {}
    for p in merged.values():
        loc = p.get("in")
        if v2 and loc == "body":
            body = {"required": bool(p.get("required")), "content_types": consumes,
                    "schema": p.get("schema", {})}
        elif v2 and loc == "formData":
            form_props[p["name"]] = {k: v for k, v in p.items() if k not in ("name", "in", "required")}
        else:
            schema = p.get("schema") or {k: p[k] for k in ("type", "format", "enum", "items") if k in p}
            params.append(Param(name=p.get("name", ""), location=loc or "query",
                                required=bool(p.get("required")) or loc == "path", schema=schema))
    if form_props:
        body = {"required": False, "content_types": consumes,
                "schema": {"type": "object", "properties": form_props}}
    return params, body


def _body_v3(rb):
    if not isinstance(rb, dict):
        return None
    content = rb.get("content") or {}
    first = next(iter(content.values()), {}) or {}
    return {"required": bool(rb.get("required")), "content_types": list(content),
            "schema": first.get("schema", {})}


def _responses(resps, v2, produces):
    out = {}
    for status, resp in (resps or {}).items():
        if not isinstance(resp, dict):
            continue
        if v2:
            schema = resp.get("schema", {})
        else:
            content = resp.get("content") or {}
            schema = (next(iter(content.values()), {}) or {}).get("schema", {})
        out[str(status)] = {"description": resp.get("description", ""), "schema": schema}
    return out


def _servers(doc, v2):
    if v2:
        host = doc.get("host")
        if not host:
            return []
        return [f"{s}://{host}{doc.get('basePath', '')}" for s in doc.get("schemes", ["https"])]
    return [s.get("url", "") for s in doc.get("servers", []) if isinstance(s, dict)]
